In [1]:
import pandas as pd

import numpy as np

import seaborn as sns

import matplotlib.pyplot as plt

import lazypredict

train_df = pd.read_csv("C:/Users/pfali/Downloads/house-prices-advanced-regression-techniques/train.csv")
test_df = pd.read_csv("C:/Users/pfali/Downloads/house-prices-advanced-regression-techniques/test.csv")

#preview data

pd.set_option('display.max_columns', None)
train_df.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities LotConfig LandSlope Neighborhood Condition1  \
0         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
1         Lvl    AllPub       FR2       Gtl      Veenker      Feedr   
2         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
3         Lvl    AllPub    Corner       Gtl      Crawfor       Norm   
4         Lvl    AllPub       FR2       Gtl      NoRidge       Norm   

  Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
0       Norm     1Fam     2Story            7            5       2003   
1       Norm     1Fam     1Story            6            8       1976   
2       Norm     1Fam     2Story            7            5       2001   
3       Norm     1Fam     2Story            7            5       1915   
4       Norm     1Fam     2Story            8            5       2000   

   YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  \
0          2003     Gable  CompShg     VinylSd     VinylSd    BrkFace   
1          1976     Gable  CompShg     MetalSd     MetalSd        NaN   
2          2002     Gable  CompShg     VinylSd     VinylSd    BrkFace   
3          1970     Gable  CompShg     Wd Sdng     Wd Shng        NaN   
4          2000     Gable  CompShg     VinylSd     VinylSd    BrkFace   

   MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure  \
0       196.0        Gd        TA      PConc       Gd       TA           No   
1         0.0        TA        TA     CBlock       Gd       TA           Gd   
2       162.0        Gd        TA      PConc       Gd       TA           Mn   
3         0.0        TA        TA     BrkTil       TA       Gd           No   
4       350.0        Gd        TA      PConc       Gd       TA           Av   

  BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  \
0          GLQ         706          Unf           0        150          856   
1          ALQ         978          Unf           0        284         1262   
2          GLQ         486          Unf           0        434          920   
3          ALQ         216          Unf           0        540          756   
4          GLQ         655          Unf           0        490         1145   

  Heating HeatingQC CentralAir Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  \
0    GasA        Ex          Y      SBrkr       856       854             0   
1    GasA        Ex          Y      SBrkr      1262         0             0   
2    GasA        Ex          Y      SBrkr       920       866             0   
3    GasA        Gd          Y      SBrkr       961       756             0   
4    GasA        Ex          Y      SBrkr      1145      1053             0   

   GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  \
0       1710             1             0         2         1             3   
1       1262             0             1         2         0             3   
2       1786             1             0         2         1             3   
3       1717             1             0         1         0             3   
4       2198             1             0         2         1             4   

   KitchenAbvGr KitchenQual  TotRmsAbvGrd Functional  Fireplaces FireplaceQu  \
0             1          Gd             8        Typ           0         NaN   
1             1          TA             6        Typ           1          TA   
2             1          Gd             6        Typ           1          TA   
3             1          Gd            

Describe data

In [2]:
train_df.describe(include='all')

Id   MSSubClass MSZoning  LotFrontage        LotArea Street  \
count   1460.000000  1460.000000     1460  1201.000000    1460.000000   1460   
unique          NaN          NaN        5          NaN            NaN      2   
top             NaN          NaN       RL          NaN            NaN   Pave   
freq            NaN          NaN     1151          NaN            NaN   1454   
mean     730.500000    56.897260      NaN    70.049958   10516.828082    NaN   
std      421.610009    42.300571      NaN    24.284752    9981.264932    NaN   
min        1.000000    20.000000      NaN    21.000000    1300.000000    NaN   
25%      365.750000    20.000000      NaN    59.000000    7553.500000    NaN   
50%      730.500000    50.000000      NaN    69.000000    9478.500000    NaN   
75%     1095.250000    70.000000      NaN    80.000000   11601.500000    NaN   
max     1460.000000   190.000000      NaN   313.000000  215245.000000    NaN   

       Alley LotShape LandContour Utilities LotConfig LandSlope Neighborhood  \
count     91     1460        1460      1460      1460      1460         1460   
unique     2        4           4         2         5         3           25   
top     Grvl      Reg         Lvl    AllPub    Inside       Gtl        NAmes   
freq      50      925        1311      1459      1052      1382          225   
mean     NaN      NaN         NaN       NaN       NaN       NaN          NaN   
std      NaN      NaN         NaN       NaN       NaN       NaN          NaN   
min      NaN      NaN         NaN       NaN       NaN       NaN          NaN   
25%      NaN      NaN         NaN       NaN       NaN       NaN          NaN   
50%      NaN      NaN         NaN       NaN       NaN       NaN          NaN   
75%      NaN      NaN         NaN       NaN       NaN       NaN          NaN   
max      NaN      NaN         NaN       NaN       NaN       NaN          NaN   

       Condition1 Condition2 BldgType HouseStyle  OverallQual  OverallCond  \
count        1460       1460     1460       1460  1460.000000  1460.000000   
unique          9          8        5          8          NaN          NaN   
top          Norm       Norm     1Fam     1Story          NaN          NaN   
freq         1260       1445     1220        726          NaN          NaN   
mean          NaN        NaN      NaN        NaN     6.099315     5.575342   
std           NaN        NaN      NaN        NaN     1.382997     1.112799   
min           NaN        NaN      NaN        NaN     1.000000     1.000000   
25%           NaN        NaN      NaN        NaN     5.000000     5.000000   
50%           NaN        NaN      NaN        NaN     6.000000     5.000000   
75%           NaN        NaN      NaN        NaN     7.000000     6.000000   
max           NaN        NaN      NaN        NaN    10.000000     9.000000   

          YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd  \
count   1460.000000   1460.000000      1460     1460        1460        1460   
unique          NaN           NaN         6        8          15          16   
top             NaN           NaN     Gable  CompShg     VinylSd     VinylSd   
freq            NaN           NaN      1141     1434         515         504   
mean    1971.267808   1984.865753       NaN      NaN         NaN         NaN   
std       30.202904     20.645407       NaN      NaN         NaN         NaN   
min     1872.000000   1950.000000       NaN      NaN         NaN         NaN   
25%     1954.000000   1967.000000       NaN      NaN         NaN         NaN   
50%     1973.000000   1994.000000       NaN      NaN         NaN         NaN   
75%     2000.000000   2004.000000       NaN      NaN         NaN         NaN   
max     2010.000000   2010.000000       NaN      NaN         NaN         NaN   

       MasVnrType   MasVnrArea ExterQual ExterCond Foundation BsmtQual  \
count         588  1452.000000      1460      1460       1460     1423   
unique          3          NaN         4         

Check nulls

In [3]:
pd.set_option('display.max_rows', None)
train_df.isnull().sum()

Id                  0
MSSubClass          0
MSZoning            0
LotFrontage       259
LotArea             0
Street              0
Alley            1369
LotShape            0
LandContour         0
Utilities           0
LotConfig           0
LandSlope           0
Neighborhood        0
Condition1          0
Condition2          0
BldgType            0
HouseStyle          0
OverallQual         0
OverallCond         0
YearBuilt           0
YearRemodAdd        0
RoofStyle           0
RoofMatl            0
Exterior1st         0
Exterior2nd         0
MasVnrType        872
MasVnrArea          8
ExterQual           0
ExterCond           0
Foundation          0
BsmtQual           37
BsmtCond           37
BsmtExposure       38
BsmtFinType1       37
BsmtFinSF1          0
BsmtFinType2       38
BsmtFinSF2          0
BsmtUnfSF           0
TotalBsmtSF         0
Heating             0
HeatingQC           0
CentralAir          0
Electrical          1
1stFlrSF            0
2ndFlrSF            0
LowQualFin

In [4]:
test_df.isnull().sum()

Id                  0
MSSubClass          0
MSZoning            4
LotFrontage       227
LotArea             0
Street              0
Alley            1352
LotShape            0
LandContour         0
Utilities           2
LotConfig           0
LandSlope           0
Neighborhood        0
Condition1          0
Condition2          0
BldgType            0
HouseStyle          0
OverallQual         0
OverallCond         0
YearBuilt           0
YearRemodAdd        0
RoofStyle           0
RoofMatl            0
Exterior1st         1
Exterior2nd         1
MasVnrType        894
MasVnrArea         15
ExterQual           0
ExterCond           0
Foundation          0
BsmtQual           44
BsmtCond           45
BsmtExposure       44
BsmtFinType1       42
BsmtFinSF1          1
BsmtFinType2       42
BsmtFinSF2          1
BsmtUnfSF           1
TotalBsmtSF         1
Heating             0
HeatingQC           0
CentralAir          0
Electrical          0
1stFlrSF            0
2ndFlrSF            0
LowQualFin

We have 83 features of various data type. Null values need to be completed. At first glance almost all features can contribute to the target, thus, initially we won't
exclude any feature.

Completing null values

MsZoning - this is nominal data, with 5 values and only 4 missing, which will be imputed with mode

LotFrontage - this is numerical, ratio, continuous data

Alley - nominal data, with two values: Pave, Grvl; 1369 out of 1460 values missing? - might be not missing but 'not applicable', let's try to  leave the feature first with transforming it to categorical data type

In [5]:
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer
from sklearn import preprocessing

train_df['MSZoning'].fillna(train_df['MSZoning'].mode()[0],inplace=True)

imp = IterativeImputer(max_iter=10, random_state = 0)

train_df['LotFrontage'] = imp.fit_transform(np.array(train_df['LotFrontage']).reshape(-1,1))
test_df['LotFrontage'] = imp.fit_transform(np.array(test_df['LotFrontage']).reshape(-1,1))

train_df['Alley'].fillna("Not",inplace=True)
test_df['Alley'].fillna("Not",inplace=True)

le = preprocessing.LabelEncoder()
train_df['Alley'] = le.fit_transform(train_df['Alley'])
test_df['Alley'] = le.fit_transform(test_df['Alley'])

train_df.head()


Id  MSSubClass MSZoning  LotFrontage  LotArea Street  Alley LotShape  \
0   1          60       RL         65.0     8450   Pave      1      Reg   
1   2          20       RL         80.0     9600   Pave      1      Reg   
2   3          60       RL         68.0    11250   Pave      1      IR1   
3   4          70       RL         60.0     9550   Pave      1      IR1   
4   5          60       RL         84.0    14260   Pave      1      IR1   

  LandContour Utilities LotConfig LandSlope Neighborhood Condition1  \
0         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
1         Lvl    AllPub       FR2       Gtl      Veenker      Feedr   
2         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
3         Lvl    AllPub    Corner       Gtl      Crawfor       Norm   
4         Lvl    AllPub       FR2       Gtl      NoRidge       Norm   

  Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
0       Norm     1Fam     2Story            7            5       2003   
1       Norm     1Fam     1Story            6            8       1976   
2       Norm     1Fam     2Story            7            5       2001   
3       Norm     1Fam     2Story            7            5       1915   
4       Norm     1Fam     2Story            8            5       2000   

   YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  \
0          2003     Gable  CompShg     VinylSd     VinylSd    BrkFace   
1          1976     Gable  CompShg     MetalSd     MetalSd        NaN   
2          2002     Gable  CompShg     VinylSd     VinylSd    BrkFace   
3          1970     Gable  CompShg     Wd Sdng     Wd Shng        NaN   
4          2000     Gable  CompShg     VinylSd     VinylSd    BrkFace   

   MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure  \
0       196.0        Gd        TA      PConc       Gd       TA           No   
1         0.0        TA        TA     CBlock       Gd       TA           Gd   
2       162.0        Gd        TA      PConc       Gd       TA           Mn   
3         0.0        TA        TA     BrkTil       TA       Gd           No   
4       350.0        Gd        TA      PConc       Gd       TA           Av   

  BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  \
0          GLQ         706          Unf           0        150          856   
1          ALQ         978          Unf           0        284         1262   
2          GLQ         486          Unf           0        434          920   
3          ALQ         216          Unf           0        540          756   
4          GLQ         655          Unf           0        490         1145   

  Heating HeatingQC CentralAir Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  \
0    GasA        Ex          Y      SBrkr       856       854             0   
1    GasA        Ex          Y      SBrkr      1262         0             0   
2    GasA        Ex          Y      SBrkr       920       866             0   
3    GasA        Gd          Y      SBrkr       961       756             0   
4    GasA        Ex          Y      SBrkr      1145      1053             0   

   GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  \
0       1710             1             0         2         1             3   
1       1262             0             1         2         0             3   
2       1786             1             0         2         1             3   
3       1717             1             0         1         0             3   
4       2198             1             0         2         1             4   

   KitchenAbvGr KitchenQual  TotRmsAbvGrd Functional  Fireplaces FireplaceQu  \
0             1          Gd             8        Typ           0         NaN   
1             1          TA             6        Typ           1          TA   
2             1          Gd             6        Typ           1          TA   
3             1          Gd      

Exterior1st and Exterior2nd - 1 missing value per each in the test dataset, indicating 'not applicable' feature; will be replaced with missing and then transformed to category.

In [6]:
test_df['Exterior1st'].fillna("Missing",inplace=True)
test_df['Exterior2nd'].fillna("Missing",inplace=True)

MasVnrType and MasVnrArea - None in MasVnrType in the training dataset means absence, which is indicated by 0 in MasVnrArea; therefore, we have in fact only 8 missing values in both features, thus, they will be imputed with mode

In [7]:
train_df['MasVnrType'].fillna(train_df['MasVnrType'].mode()[0],inplace=True)
train_df['MasVnrArea'].fillna(train_df['MasVnrArea'].mode()[0],inplace=True)

MasVnrType and MasVnrArea - in the test dataset 15 null values of MasVnrArea indicate 'not applicable', so to be replaced with zeros. However, MasVnrType have more missing values, thus, possibly it will be better to use KNNImputer to impute nulls.

In [8]:
test_df['MasVnrArea'].fillna(0,inplace=True)

from sklearn import preprocessing
from sklearn.impute import KNNImputer

imputer = KNNImputer(n_neighbors=1460)
le = preprocessing.LabelEncoder()

test_df['MasVnrType'] = le.fit_transform(test_df['MasVnrType'])
test_df['MasVnrType'] = imputer.fit_transform(test_df[['MasVnrType']]).ravel()

BsmtQual, BsmtCond, BsmtFinType1 have 37 missing values each which equals to the number of 0-value data in TotalBsmtSF, which means missing values indicate no basement. Thus, can be filled with "Missing". In the test dataset it is similar - 42 zeros in TotalBsmtSF against 44, 45, 44, 42 and 42 nulls in BsmtQual, BsmtCond, BsmtFinType1, BsmtExposure, BsmtFinType2, so all will be imputed with "Missing" and then transformed to categories.

In [9]:
train_df['BsmtQual'].fillna("Missing",inplace=True)
train_df['BsmtCond'].fillna("Missing",inplace=True)
train_df['BsmtFinType1'].fillna("Missing",inplace=True)

test_df['BsmtQual'].fillna("Missing",inplace=True)
test_df['BsmtCond'].fillna("Missing",inplace=True)
test_df['BsmtFinType1'].fillna("Missing",inplace=True)
test_df['BsmtExposure'].fillna("Missing",inplace=True)
test_df['BsmtFinType2'].fillna("Missing",inplace=True)

BsmtExposure, BsmtFinType2 - both have 1 null value in the training dataset, but seems to be dependent of other basement-describing features, thus, to be replaced using KNNImputer
BsmtUnfSF, TotalBsmtSF - both have 1 null value in the test dataset, but seems to be dependent of other basement-describing features, thus, to be replaced using KNNImputer

In [10]:
imputer = KNNImputer(n_neighbors=1460)
le = preprocessing.LabelEncoder()

train_df['BsmtExposure'] = le.fit_transform(train_df['BsmtExposure'])
train_df['BsmtFinType2'] = le.fit_transform(train_df['BsmtFinType2'])

train_df['BsmtExposure'] = imputer.fit_transform(train_df[['BsmtExposure']]).ravel()
train_df['BsmtFinType2'] = imputer.fit_transform(train_df[['BsmtFinType2']]).ravel()

test_df['BsmtUnfSF'] = le.fit_transform(test_df['BsmtUnfSF'])
test_df['TotalBsmtSF'] = le.fit_transform(test_df['TotalBsmtSF'])

test_df['BsmtUnfSF'] = imputer.fit_transform(test_df[['BsmtUnfSF']]).ravel()
test_df['TotalBsmtSF'] = imputer.fit_transform(test_df[['TotalBsmtSF']]).ravel()

Electrical - this is nominal data with 1 null value; to be imputed with mode
BsmtFullBath, BsmtHalfBath, Functional - these are ratio, continuous data with 1 or 2 null values; to be imputed with mode

In [11]:
train_df['Electrical'].fillna(train_df['Electrical'].mode()[0],inplace=True)

test_df['BsmtFullBath'].fillna(test_df['BsmtFullBath'].mode()[0],inplace=True)
test_df['BsmtHalfBath'].fillna(test_df['BsmtHalfBath'].mode()[0],inplace=True)
test_df['Functional'].fillna(test_df['Functional'].mode()[0],inplace=True)

Missing values for FireplaceQu mean in fact 'not applicable' feature (=no fireplace), thus, can be replaced with 0. The same for features related to garages: GarageType, GarageYrBlt, GarageFinish, GarageQual, GarageCond. These applies to both the test and training dataset. The same for PoolQC - nulls indicate no pool, but only in the training dataset.

In [12]:
train_df['FireplaceQu'].fillna("Missing",inplace=True)
train_df['GarageType'].fillna("Missing",inplace=True)
train_df['GarageYrBlt'].fillna(0,inplace=True)
train_df['GarageFinish'].fillna("Missing",inplace=True)
train_df['GarageQual'].fillna("Missing",inplace=True)
train_df['GarageCond'].fillna("Missing",inplace=True)
train_df['PoolQC'].fillna("Missing",inplace=True)

test_df['FireplaceQu'].fillna("Missing",inplace=True)
test_df['GarageType'].fillna("Missing",inplace=True)
test_df['GarageYrBlt'].fillna(0,inplace=True)
test_df['GarageFinish'].fillna("Missing",inplace=True)
test_df['GarageQual'].fillna("Missing",inplace=True)
test_df['GarageCond'].fillna("Missing",inplace=True)

Fence and MiscFeature - both have many missing values, which can be either nulls or 'not applicable' values. Therefore, both will be imputed with KNNImputer. Additionally, PoolQC will be imputed the same way in the test dataset.
The same way GarageCars and GarageArea in the test dataset will be imputed, because they can be correlated with other car-type features.

In [13]:
train_df['Fence'] = le.fit_transform(train_df['Fence'])
train_df['MiscFeature'] = le.fit_transform(train_df['MiscFeature'])
train_df['Fence'] = imputer.fit_transform(train_df[['Fence']]).ravel()
train_df['MiscFeature'] = imputer.fit_transform(train_df[['MiscFeature']]).ravel()

test_df['Fence'] = le.fit_transform(test_df['Fence'])
test_df['MiscFeature'] = le.fit_transform(test_df['MiscFeature'])
test_df['Fence'] = imputer.fit_transform(test_df[['Fence']]).ravel()
test_df['MiscFeature'] = imputer.fit_transform(test_df[['MiscFeature']]).ravel()
test_df['PoolQC'] = le.fit_transform(test_df['PoolQC'])
test_df['PoolQC'] = imputer.fit_transform(test_df[['PoolQC']]).ravel()

test_df['GarageCars'] = le.fit_transform(test_df['GarageCars'])
test_df['GarageArea'] = le.fit_transform(test_df['GarageArea'])
test_df['GarageCars'] = imputer.fit_transform(test_df[['GarageCars']]).ravel()
test_df['GarageArea'] = imputer.fit_transform(test_df[['GarageArea']]).ravel()

SaleType - this is nominal data with 1 value missing in the test dataset, thus, to be replaced with mode.

In [14]:
test_df['SaleType'].fillna(test_df['SaleType'].mode()[0],inplace=True)

In [15]:
train_df.isnull().sum()

Id               0
MSSubClass       0
MSZoning         0
LotFrontage      0
LotArea          0
Street           0
Alley            0
LotShape         0
LandContour      0
Utilities        0
LotConfig        0
LandSlope        0
Neighborhood     0
Condition1       0
Condition2       0
BldgType         0
HouseStyle       0
OverallQual      0
OverallCond      0
YearBuilt        0
YearRemodAdd     0
RoofStyle        0
RoofMatl         0
Exterior1st      0
Exterior2nd      0
MasVnrType       0
MasVnrArea       0
ExterQual        0
ExterCond        0
Foundation       0
BsmtQual         0
BsmtCond         0
BsmtExposure     0
BsmtFinType1     0
BsmtFinSF1       0
BsmtFinType2     0
BsmtFinSF2       0
BsmtUnfSF        0
TotalBsmtSF      0
Heating          0
HeatingQC        0
CentralAir       0
Electrical       0
1stFlrSF         0
2ndFlrSF         0
LowQualFinSF     0
GrLivArea        0
BsmtFullBath     0
BsmtHalfBath     0
FullBath         0
HalfBath         0
BedroomAbvGr     0
KitchenAbvGr

Check data types of features to possible replace nominal features with numerical categories

In [16]:
print(train_df.dtypes)

Id                 int64
MSSubClass         int64
MSZoning          object
LotFrontage      float64
LotArea            int64
Street            object
Alley              int32
LotShape          object
LandContour       object
Utilities         object
LotConfig         object
LandSlope         object
Neighborhood      object
Condition1        object
Condition2        object
BldgType          object
HouseStyle        object
OverallQual        int64
OverallCond        int64
YearBuilt          int64
YearRemodAdd       int64
RoofStyle         object
RoofMatl          object
Exterior1st       object
Exterior2nd       object
MasVnrType        object
MasVnrArea       float64
ExterQual         object
ExterCond         object
Foundation        object
BsmtQual          object
BsmtCond          object
BsmtExposure     float64
BsmtFinType1      object
BsmtFinSF1         int64
BsmtFinType2     float64
BsmtFinSF2         int64
BsmtUnfSF          int64
TotalBsmtSF        int64
Heating           object


In [17]:
print(test_df.dtypes)

Id                 int64
MSSubClass         int64
MSZoning          object
LotFrontage      float64
LotArea            int64
Street            object
Alley              int32
LotShape          object
LandContour       object
Utilities         object
LotConfig         object
LandSlope         object
Neighborhood      object
Condition1        object
Condition2        object
BldgType          object
HouseStyle        object
OverallQual        int64
OverallCond        int64
YearBuilt          int64
YearRemodAdd       int64
RoofStyle         object
RoofMatl          object
Exterior1st       object
Exterior2nd       object
MasVnrType       float64
MasVnrArea       float64
ExterQual         object
ExterCond         object
Foundation        object
BsmtQual          object
BsmtCond          object
BsmtExposure      object
BsmtFinType1      object
BsmtFinSF1       float64
BsmtFinType2      object
BsmtFinSF2       float64
BsmtUnfSF        float64
TotalBsmtSF      float64
Heating           object


In [18]:
train_df['MSZoning'] = le.fit_transform(train_df['MSZoning'])
train_df['Street'] = le.fit_transform(train_df['Street'])
train_df['LotShape'] = le.fit_transform(train_df['LotShape'])
train_df['LandContour'] = le.fit_transform(train_df['LandContour'])
train_df['Utilities'] = le.fit_transform(train_df['Utilities'])
train_df['LotConfig'] = le.fit_transform(train_df['LotConfig'])
train_df['LandSlope'] = le.fit_transform(train_df['LandSlope'])
train_df['Neighborhood'] = le.fit_transform(train_df['Neighborhood'])
train_df['Condition1'] = le.fit_transform(train_df['Condition1'])
train_df['Condition2'] = le.fit_transform(train_df['Condition2'])
train_df['BldgType'] = le.fit_transform(train_df['BldgType'])
train_df['HouseStyle'] = le.fit_transform(train_df['HouseStyle'])
train_df['RoofStyle'] = le.fit_transform(train_df['RoofStyle'])
train_df['RoofMatl'] = le.fit_transform(train_df['RoofMatl'])
train_df['Exterior1st'] = le.fit_transform(train_df['Exterior1st'])
train_df['Exterior2nd'] = le.fit_transform(train_df['Exterior2nd'])
train_df['MasVnrType'] = le.fit_transform(train_df['MasVnrType'])
train_df['ExterQual'] = le.fit_transform(train_df['ExterQual'])
train_df['ExterCond'] = le.fit_transform(train_df['ExterCond'])
train_df['Foundation'] = le.fit_transform(train_df['Foundation'])
train_df['BsmtQual'] = le.fit_transform(train_df['BsmtQual'])
train_df['BsmtCond'] = le.fit_transform(train_df['BsmtCond'])
train_df['BsmtFinType1'] = le.fit_transform(train_df['BsmtFinType1'])
train_df['Heating'] = le.fit_transform(train_df['Heating'])
train_df['HeatingQC'] = le.fit_transform(train_df['HeatingQC'])
train_df['CentralAir'] = le.fit_transform(train_df['CentralAir'])
train_df['Electrical'] = le.fit_transform(train_df['Electrical'])
train_df['KitchenQual'] = le.fit_transform(train_df['KitchenQual'])
train_df['Functional'] = le.fit_transform(train_df['Functional'])
train_df['FireplaceQu'] = le.fit_transform(train_df['FireplaceQu'])
train_df['GarageType'] = le.fit_transform(train_df['GarageType'])
train_df['GarageFinish'] = le.fit_transform(train_df['GarageFinish'])
train_df['GarageQual'] = le.fit_transform(train_df['GarageQual'])
train_df['GarageYrBlt'] = le.fit_transform(train_df['GarageYrBlt'])
train_df['GarageCond'] = le.fit_transform(train_df['GarageCond'])
train_df['PavedDrive'] = le.fit_transform(train_df['PavedDrive'])
train_df['PoolQC'] = le.fit_transform(train_df['PoolQC'])
train_df['SaleType'] = le.fit_transform(train_df['SaleType'])
train_df['SaleCondition'] = le.fit_transform(train_df['SaleCondition'])


test_df['MSZoning'] = le.fit_transform(test_df['MSZoning'])
test_df['Street'] = le.fit_transform(test_df['Street'])
test_df['LotShape'] = le.fit_transform(test_df['LotShape'])
test_df['LandContour'] = le.fit_transform(test_df['LandContour'])
test_df['Utilities'] = le.fit_transform(test_df['Utilities'])
test_df['LotConfig'] = le.fit_transform(test_df['LotConfig'])
test_df['LandSlope'] = le.fit_transform(test_df['LandSlope'])
test_df['Neighborhood'] = le.fit_transform(test_df['Neighborhood'])
test_df['Condition1'] = le.fit_transform(test_df['Condition1'])
test_df['Condition2'] = le.fit_transform(test_df['Condition2'])
test_df['BldgType'] = le.fit_transform(test_df['BldgType'])
test_df['HouseStyle'] = le.fit_transform(test_df['HouseStyle'])
test_df['RoofStyle'] = le.fit_transform(test_df['RoofStyle'])
test_df['RoofMatl'] = le.fit_transform(test_df['RoofMatl'])
test_df['Exterior1st'] = le.fit_transform(test_df['Exterior1st'])
test_df['Exterior2nd'] = le.fit_transform(test_df['Exterior2nd'])
test_df['MasVnrType'] = le.fit_transform(test_df['MasVnrType'])
test_df['ExterQual'] = le.fit_transform(test_df['ExterQual'])
test_df['ExterCond'] = le.fit_transform(test_df['ExterCond'])
test_df['Foundation'] = le.fit_transform(test_df['Foundation'])
test_df['BsmtQual'] = le.fit_transform(test_df['BsmtQual'])
test_df['BsmtCond'] = le.fit_transform(test_df['BsmtCond'])
test_df['BsmtExposure'] = le.fit_transform(test_df['BsmtExposure'])
test_df['BsmtFinType1'] = le.fit_transform(test_df['BsmtFinType1'])
test_df['BsmtFinType2'] = le.fit_transform(test_df['BsmtFinType2'])
test_df['Heating'] = le.fit_transform(test_df['Heating'])
test_df['HeatingQC'] = le.fit_transform(test_df['HeatingQC'])
test_df['CentralAir'] = le.fit_transform(test_df['CentralAir'])
test_df['Electrical'] = le.fit_transform(test_df['Electrical'])
test_df['KitchenQual'] = le.fit_transform(test_df['KitchenQual'])
test_df['Functional'] = le.fit_transform(test_df['Functional'])
test_df['FireplaceQu'] = le.fit_transform(test_df['FireplaceQu'])
test_df['GarageType'] = le.fit_transform(test_df['GarageType'])
test_df['GarageFinish'] = le.fit_transform(test_df['GarageFinish'])
test_df['GarageQual'] = le.fit_transform(test_df['GarageQual'])
test_df['GarageCond'] = le.fit_transform(test_df['GarageCond'])
test_df['PavedDrive'] = le.fit_transform(test_df['PavedDrive'])
test_df['SaleType'] = le.fit_transform(test_df['SaleType'])
test_df['SaleCondition'] = le.fit_transform(test_df['SaleCondition'])


In [19]:
print(train_df.dtypes)

Id                 int64
MSSubClass         int64
MSZoning           int32
LotFrontage      float64
LotArea            int64
Street             int32
Alley              int32
LotShape           int32
LandContour        int32
Utilities          int32
LotConfig          int32
LandSlope          int32
Neighborhood       int32
Condition1         int32
Condition2         int32
BldgType           int32
HouseStyle         int32
OverallQual        int64
OverallCond        int64
YearBuilt          int64
YearRemodAdd       int64
RoofStyle          int32
RoofMatl           int32
Exterior1st        int32
Exterior2nd        int32
MasVnrType         int32
MasVnrArea       float64
ExterQual          int32
ExterCond          int32
Foundation         int32
BsmtQual           int32
BsmtCond           int32
BsmtExposure     float64
BsmtFinType1       int32
BsmtFinSF1         int64
BsmtFinType2     float64
BsmtFinSF2         int64
BsmtUnfSF          int64
TotalBsmtSF        int64
Heating            int32


In [20]:
train_df.describe(include='all')

Id   MSSubClass     MSZoning  LotFrontage        LotArea  \
count  1460.000000  1460.000000  1460.000000  1460.000000    1460.000000   
mean    730.500000    56.897260     3.028767    70.049958   10516.828082   
std     421.610009    42.300571     0.632017    22.024023    9981.264932   
min       1.000000    20.000000     0.000000    21.000000    1300.000000   
25%     365.750000    20.000000     3.000000    60.000000    7553.500000   
50%     730.500000    50.000000     3.000000    70.049958    9478.500000   
75%    1095.250000    70.000000     3.000000    79.000000   11601.500000   
max    1460.000000   190.000000     4.000000   313.000000  215245.000000   

            Street        Alley     LotShape  LandContour    Utilities  \
count  1460.000000  1460.000000  1460.000000  1460.000000  1460.000000   
mean      0.995890     0.993836     1.942466     2.777397     0.000685   
std       0.063996     0.249667     1.409156     0.707666     0.026171   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       1.000000     1.000000     0.000000     3.000000     0.000000   
50%       1.000000     1.000000     3.000000     3.000000     0.000000   
75%       1.000000     1.000000     3.000000     3.000000     0.000000   
max       1.000000     2.000000     3.000000     3.000000     1.000000   

         LotConfig    LandSlope  Neighborhood   Condition1   Condition2  \
count  1460.000000  1460.000000   1460.000000  1460.000000  1460.000000   
mean      3.019178     0.062329     12.251370     2.031507     2.008219   
std       1.622634     0.276232      6.013735     0.868515     0.259040   
min       0.000000     0.000000      0.000000     0.000000     0.000000   
25%       2.000000     0.000000      7.000000     2.000000     2.000000   
50%       4.000000     0.000000     12.000000     2.000000     2.000000   
75%       4.000000     0.000000     17.000000     2.000000     2.000000   
max       4.000000     2.000000     24.000000     8.000000     7.000000   

          BldgType   HouseStyle  OverallQual  OverallCond    YearBuilt  \
count  1460.000000  1460.000000  1460.000000  1460.000000  1460.000000   
mean      0.493151     3.038356     6.099315     5.575342  1971.267808   
std       1.198277     1.911305     1.382997     1.112799    30.202904   
min       0.000000     0.000000     1.000000     1.000000  1872.000000   
25%       0.000000     2.000000     5.000000     5.000000  1954.000000   
50%       0.000000     2.000000     6.000000     5.000000  1973.000000   
75%       0.000000     5.000000     7.000000     6.000000  2000.000000   
max       4.000000     7.000000    10.000000     9.000000  2010.000000   

       YearRemodAdd    RoofStyle     RoofMatl  Exterior1st  Exterior2nd  \
count   1460.000000  1460.000000  1460.000000  1460.000000  1460.000000   
mean    1984.865753     1.410274     1.075342     9.624658    10.339726   
std       20.645407     0.834998     0.599127     3.197659     3.540570   
min     1950.000000     0.000000     0.000000     0.000000     0.000000   
25%     1967.000000     1.000000     1.000000     8.000000     8.000000   
50%     1994.000000     1.000000     1.000000    12.000000    13.000000   
75%     2004.000000     1.000000     1.000000    12.000000    13.000000   
max     2010.000000     5.000000     7.000000    14.000000    15.000000   

        MasVnrType   MasVnrArea    ExterQual    ExterCond   Foundation  \
count  1460.000000  1460.000000  1460.000000  1460.000000  1460.000000   
mean      1.077397   103.117123     2.539726     3.733562     1.396575   
std       0.303345   180.731373     0.693995     0.731807     0.722394   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       1.000000     0.000000     2.000000     4.000000     1.000000   
50%       1.000000     0.000000     3.000000     4.000000     1.000000   
75%       1.000000   164.250000     3.000000     4.000000     2.000000   
max       2.000000  1600.000000     3.000000     4.00

Prepare correlation matrix to see which features correlate most with the target to possible use it later

In [21]:
corr_matrix = train_df.corr()
corr_matrix["SalePrice"].sort_values(ascending=False)

SalePrice        1.000000
OverallQual      0.790982
GrLivArea        0.708624
GarageCars       0.640409
GarageArea       0.623431
TotalBsmtSF      0.613581
1stFlrSF         0.605852
FullBath         0.560664
TotRmsAbvGrd     0.533723
GarageYrBlt      0.527527
YearBuilt        0.522897
YearRemodAdd     0.507101
MasVnrArea       0.472614
Fireplaces       0.466929
BsmtFinSF1       0.386420
Foundation       0.382479
LotFrontage      0.334901
WoodDeckSF       0.324413
MasVnrType       0.322979
2ndFlrSF         0.319334
OpenPorchSF      0.315856
HalfBath         0.284108
LotArea          0.263843
CentralAir       0.251328
GarageCond       0.246705
Electrical       0.234945
PavedDrive       0.231357
BsmtFullBath     0.227122
RoofStyle        0.222405
BsmtUnfSF        0.214479
SaleCondition    0.213092
Neighborhood     0.210851
GarageQual       0.205963
HouseStyle       0.180163
BedroomAbvGr     0.168213
Fence            0.140640
RoofMatl         0.132383
ExterCond        0.117303
Functional  

Split train data into train set and validation set

In [22]:
from sklearn.model_selection import train_test_split

X = train_df.drop(["SalePrice"], axis = 1)
y = train_df["SalePrice"]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)

Check how various models perform

In [23]:
from lazypredict.Supervised import LazyRegressor

reg = LazyRegressor(verbose=0, ignore_warnings=False, custom_metric=None)
models_pred = reg.fit(X_train, X_test, y_train, y_test)
print(models_pred)

 71%|██████████████████████████████████████████████████████████▌                       | 30/42 [00:09<00:02,  4.69it/s]

QuantileRegressor model failed to execute
Solver interior-point is not anymore available in SciPy >= 1.11.0.


100%|██████████████████████████████████████████████████████████████████████████████████| 42/42 [00:15<00:00,  2.64it/s]

[LightGBM] [Warning] Auto-choosing row-wise multi-threading, the overhead of testing was 0.000418 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 3366
[LightGBM] [Info] Number of data points in the train set: 1095, number of used features: 72
[LightGBM] [Info] Start training from score 182319.757078
(                               Adjusted R-Squared  R-Squared      RMSE  \
Model                                                                    
GradientBoostingRegressor                    0.89       0.91  24494.20   
RandomForestRegressor                        0.87       0.90  26490.60   
HistGradientBoostingRegressor                0.86       0.89  26657.56   
LGBMRegressor                                0.86       0.89  27145.56   
ExtraTreesRegressor                          0.86       0.89  27337.65   
XGBRegressor                                 0.84       0.87  29080.

GBR looks the most promising, but let's try to analyse the models again with dropping some poorly correlated to the target features.

In [25]:
from sklearn.model_selection import train_test_split

train_df.drop(["BsmtFinType2","Condition2","BsmtFinSF2","Utilities","BsmtHalfBath","MiscVal","LandContour","PoolArea","BsmtCond","Alley","MiscFeature","LandSlope","MoSold","3SsnPorch","Street","LowQualFinSF","YrSold","SaleType","LotConfig","OverallCond","MSSubClass","BldgType","FireplaceQu","BsmtFinType1","Heating"], axis = 1)

X = train_df.drop(["SalePrice"], axis = 1)
y = train_df["SalePrice"]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)


In [ ]:
reg = LazyRegressor(verbose=0, ignore_warnings=False, custom_metric=None)
models_pred = reg.fit(X_train, X_test, y_train, y_test)
print(models_pred)

In [29]:
test_df.drop(["BsmtFinType2","Condition2","BsmtFinSF2","Utilities","BsmtHalfBath","MiscVal","LandContour","PoolArea","BsmtCond","Alley","MiscFeature","LandSlope","MoSold","3SsnPorch","Street","LowQualFinSF","YrSold","SaleType","LotConfig","OverallCond","MSSubClass","BldgType","FireplaceQu","BsmtFinType1","Heating"], axis = 1)

from sklearn.ensemble import GradientBoostingRegressor

reg = GradientBoostingRegressor(random_state=0)
reg.fit(X_train, y_train)
pred=reg.predict(X_test)
score=reg.score(X_test, y_test)

print(score)

0.9144649654194527
